module/zfs/dmu_objset.c

```diff
@@ -1072,6 +1079,58 @@ dmu_objset_snapshot_one(const char *fsname, const char *snapname)
 	return (err);
 }
 
+typedef struct dmu_objset_upgrade_args {
+	objset_t *dua_os;
+	int (*dua_cb)(objset_t *);
+} dmu_objset_upgrade_arg_t;
+
+static void
+dmu_objset_upgrade_trampoline(void *data)
+{
+	dmu_objset_upgrade_arg_t *arg = data;
+	objset_t *os = arg->dua_os;
+
+	cv_signal(&os->os_upgrade_cv);
+
+	mutex_enter(&os->os_upgrade_lock);
+	if (!os->os_upgrade_exit) {
+		mutex_exit(&os->os_upgrade_lock);
+
+		(void) arg->dua_cb(os);
+		mutex_enter(&os->os_upgrade_lock);
+	}
+	os->os_upgrade_thread = NULL;
+	cv_broadcast(&os->os_upgrade_cv);
+	mutex_exit(&os->os_upgrade_lock);
```

Comment:
For this function to be safe you're going to need to take a hold on the objset like in `zfs_ioc_userspace_upgrade()`.  If you don't the objset may be freed at any time.  You should be able to use the same logic as in the `zfs_ioc_userobjspace_upgrade()` function if you pass in the full dataset name as a string instead.

This could explain the [crashes in the automated testing](http://build.zfsonlinux.org/builders/Amazon%202015.09%20x86_64%20%28TEST%29/builds/2358/steps/shell_7/logs/console).

To make error handling easier the new thread can also always exit on unexpected error.  It should be automatically restarted up until the point `dmu_objset_userobjspace_present()` is gets set.